tachyon/math/base/byinverter.h

```diff
@@ -0,0 +1,469 @@
+      result.data[i] = sum & CInt::MASK;
+      carry = sum >> B;
+    }
+    return result;
+  }
+
+  constexpr CInt operator*(const CInt &other) const {
+    CInt result;
+    for (usize i = 0; i < L; i++) {
+      u64 carry = 0;
+      for (usize k = i; k < L; k++) {
+        u128 sum =
+            (u128)this->data[i] * other.data[k - i] + result.data[k] + carry;
+        result.data[k] = (u64)(sum & CInt::MASK);
+        carry = (u64)(sum >> B);
+      }
+    }
+    return result;
+  }
+
+  CInt operator*(i64 other) const {
+    CInt result;
+    u64 mask = 0;
+    u64 carry = 0;
+    // If the short multiplicand is non-negative, the standard multiplication
+    // algorithm is performed. Otherwise, the product of the additively negated
+    // multiplicands is found as follows. Since for the two's complement code
+    // the additive negation is the result of adding 1 to the bitwise inverted
+    // argument's representation, for any encoded integers x and y we have
+    // x * y = (-x) * (-y) = (!x + 1) * (-y) = !x * (-y) + (-y),  where "!" is
+    // the bitwise inversion and arithmetic operations are performed according
+    // to the rules of the code. If the short multiplicand is negative, the
+    // algorithm below uses this formula by substituting the short multiplicand
+    // for y and turns into the modified standard multiplication algorithm,
+    // where the carry flag is initialized with the additively negated short
+    // multiplicand and the chunks of the long multiplicand are bitwise inverted
+    if (other < 0) {
+      mask = CInt::MASK;
+      carry = other = -other;
+    }
+    for (usize i = 0; i < L; i++) {
+      u128 sum = (u128)(this->data[i] ^ mask) * other + carry;
+      result.data[i] = (u64)(sum & CInt::MASK);
+      carry = (u64)(sum >> B);
+    }
+    return result;
+  }
+
+ private:
+  /// B-bit chunks representing the current number
+  u64 data[L] = {0};
+
+  /// Creates an instance representing zero
+  constexpr CInt() {}
+
+  /// Stores at the address "output" the array of O-bit chunks, which represents
+  /// the big unsigned integer equal modulo 2 ^ (O * "osize") to the input big
+  /// unsigned integer stored at the address "input" as an array of "isize"
+  /// I-bit chunks. The ordering of the chunks in these arrays is little-endian
+  template <usize I, usize O>
+  constexpr static void convert(const u64 *input, const usize isize,
+                                u64 *output, const usize osize) {
+    usize bits = 0;
+    usize total = least(isize * I, osize * O);
+    for (usize i = 0; i < osize; i++) {
+      output[i] = 0;
+    }
+    while (bits < total) {
+      usize i = bits % I;
+      usize o = bits % O;
+      output[bits / O] |= (input[bits / I] >> i) << o;
+      bits += least(I - i, O - o);
+    }
+    u64 mask = ~0UL >> (64 - O);
+    usize filled = (total + O - 1) / O;
+    while (filled > 0) {
+      filled -= 1;
+      output[filled] &= mask;
+    }
```

Comment:
```suggestion
    for (size_t i = filled - 1; i != SIZE_MAX; --i) {
      output[i] &= mask;
    }
```

Reply:
However, for people, who do not know the two's complement code, the new version will be obscure. Should I leave it as is?

Reply:
Let it be 
![image](https://github.com/user-attachments/assets/76d02392-6542-44b4-9496-257211b34f1c)